bpfd/src/bpf.rs

```diff
@@ -41,18 +41,145 @@ const MAPS_MODE: u32 = 0o0660;
+        let prog = prog.clone();
+        self.programs.insert(id, prog)
+    }
+
+    fn remove(&mut self, id: &Uuid) -> Option<Program> {
+        self.programs.remove(id)
+    }
+
+    fn get_mut(&mut self, id: &Uuid) -> Option<&mut Program> {
+        self.programs.get_mut(id)
+    }
+
+    fn get(&self, id: &Uuid) -> Option<&Program> {
+        self.programs.get(id)
+    }
+
+    fn contains_key(&self, id: &Uuid) -> bool {
+        self.programs.contains_key(id)
+    }
+
+    fn next_avail_id(&self, program: &Program) -> usize {
+        self.programs
+            .iter()
+            .filter(|(_, p)| {
+                if p.kind() == program.kind() {
+                    p.if_index() == program.if_index() && p.direction() == program.direction()
+                } else {
+                    false
+                }
+            })
+            .collect::<HashMap<_, _>>()
+            .len()
+    }
+
+    fn collect_programs(
+        &mut self,
+        program_type: ProgramType,
+        if_index: Option<u32>,
+        direction: Option<Direction>,
+    ) -> Vec<(&Uuid, &mut Program)> {
+        let mut results = vec![];
+        for (k, v) in self.programs.iter_mut() {
+            if v.kind() == program_type && v.if_index() == if_index && v.direction() == direction {
+                results.push((k, v))
+            }
+        }
+        results
+    }
+
+    fn sort_programs(
+        &mut self,
+        program_type: ProgramType,
+        if_index: Option<u32>,
+        direction: Option<Direction>,
+    ) {
+        let mut extensions = self
+            .programs
+            .iter_mut()
+            .filter_map(|(k, v)| {
+                if v.kind() == program_type {
+                    if v.if_index() == if_index && v.direction() == direction {
+                        Some((k, v))
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<(&Uuid, &mut Program)>>();
+
+        extensions.sort_by_key(|(_, b)| {
+            (
+                b.priority(),
+                b.attached().unwrap(),
+                b.data()
+                    .expect("All Bpfd programs should have ProgramData")
+                    .name()
+                    .to_owned(),
+            )
+        });
+        for (i, (_, v)) in extensions.iter_mut().enumerate() {
+            v.set_position(Some(i));
+        }
+    }
+
+    fn kernel_infos(&self) -> HashMap<u32, Program> {
+        self.programs
+            .values()
+            .map(|p| {
+                let kernel_info = p
+                    .data()
+                    .expect("All Bpfd programs should have ProgramData")
+                    .kernel_info()
+                    .expect("Loaded Bpfd programs should have kernel information");
+
+                (kernel_info.id, p.to_owned())
+            })
+            .collect()
```

Comment:
Again, if this can be an iterator we avoid calling `collect()` and leave it up to the caller what to do with it.

Reply:
Gave this a try, just had to add some manual lifetimes

Reply:
Done